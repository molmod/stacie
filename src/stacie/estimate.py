# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Algorithm to estimate the autocorrelation integral."""

import warnings

import attrs
import numpy as np
from numpy.typing import NDArray
from scipy.optimize import minimize

from .cost import LowFreqCost
from .cutoff import CutoffCriterion, underfitting_criterion
from .model import ExpTailModel, SpectrumModel
from .rpi import build_xgrid_exp, rpi_opt
from .spectrum import Spectrum

__all__ = ("Result", "FCutWarning", "estimate_acint", "fit_model_spectrum")


@attrs.define
class Result:
    """Container class holding all the results of the autocorrelation integral estimate."""

    spectrum: Spectrum = attrs.field()
    """The input spectrum from which the autocorrelation integral is estimated."""

    nfit: int = attrs.field()
    """
    The number of low-frequency spectrum data points used to fit the model.
    This value is the best among a series of tested cutoffs,
    i.e. by minimizing the cutoff criterion.
    """

    history: dict[int, dict[str]] = attrs.field()
    """History of nfit optimization.

    The key is ``nfit``, the number of frequencies fitted to.
    Each value is a dictionary returned by :func:`fit_model_spectrum`.
    """

    @property
    def props(self) -> dict[str]:
        """Properties computed from the fit up to the selected spectrum cutoff.

        This is a shortcut for ``history[nfit]``.
        See return value of :func:`fit_model_spectrum` for more details.
        """
        return self.history[self.nfit]

    @property
    def acint(self) -> float:
        """The autocorrelation integral."""
        return self.props["acint"]

    @property
    def acint_std(self) -> float:
        """The uncertainty of the autocorrelation integral."""
        return self.props["acint_std"]

    @property
    def corrtime_exp(self) -> float:
        """The exponential correlation time."""
        return self.props["corrtime"]

    @property
    def corrtime_exp_std(self) -> float:
        """The uncertainty of the exponential correlation time."""
        return self.props["corrtime_std"]

    @property
    def corrtime_int(self) -> float:
        """The integrated correlation time."""
        return self.props["acint"] / (2 * self.spectrum.prefactor * self.spectrum.variance)

    @property
    def corrtime_int_std(self) -> float:
        """The uncertainty of the integrated correlation time."""
        return self.props["acint_std"] / (2 * self.spectrum.prefactor * self.spectrum.variance)


class FCutWarning(Warning):
    """Raised when there is an issue with the frequency cutoff."""


def estimate_acint(
    spectrum: Spectrum,
    *,
    fcutmax: float | None = None,
    maxscan: int = 100,
    nfitmin: int = 10,
    nfitmax_hard: int = 1000,
    model: SpectrumModel | None = None,
    cutoff_criterion: CutoffCriterion = underfitting_criterion,
    verbose: bool = False,
) -> Result:
    """Estimate the integral of the autocorrelation function.

    It is recommended to leave the keyword arguments to their default values,
    except for methodological testing.

    This function fits a model to the low-frequency portion of the spectrum
    and derives an estimate of the autocorrelation (and its uncertainty) from the fit.
    The model is fitted to a part of the spectrum op to a cutoff frequency.
    Multiple cutoffs are tested, each resulting in another number of spectrum amplitudes in the fit,
    and the one that minimizes the ``cutoff_criterion`` is selected as the best solution.

    Parameters
    ----------
    spectrum
        The power spectrum and related metadata,
        used as inputs for the estimation of the autocorrelation integral.
        This object can be prepared with the function: :py:func:`stacie.spectrum.compute_spectrum`.
    fcutmax
        The maximum cutoff on the frequency axis (units of frequency),
        which corresponds to the largest value for ``nfit``.
    maxscan
        The maximum number of cutoffs to test.
        If 1, then only the given ``fcutmax`` is used.
    nfitmin
        The minimal amount of frequency data points to use in the fit.
    nfitmax_hard
        The maximal amount of frequency data points to use in the fit.
        This puts an upper bound on the computational cost of the fit.
        If this upper limit is stricter than that of ``fcutmax``, a warning is raised.
    model
        The model used to fit the low-frequency part of the spectrum.
        The default is an instance of :py:class:`stacie.model.ExpTailModel`.
    cutoff_criterion
        The criterion function that is minimized to find the best cutoff
        (and thus number of points included in the fit).
    verbose
        Set this to ``True`` to print progress information of the frequency cutoff search
        to the standard output.

    Returns
    -------
    result
        The inputs, intermediate results and outputs or the algorithm.
    """
    if model is None:
        model = ExpTailModel()
    history = {}

    if verbose and maxscan > 1:
        print("   nfit   criterion  incumbent")
        scratch = {}

    def compute_criterion(ifit: int):
        """Criterion to be minimized to find the best frequency cutoff.

        Parameters
        ----------
        ifit
            The index in the nfits list to get the right cutoff.
        include_failed
            When ``True``, add failed fits to the history.
        """
        nfit = nfits[ifit]
        props = fit_model_spectrum(
            spectrum.timestep,
            spectrum.freqs,
            spectrum.amplitudes,
            spectrum.ndofs,
            nfit,
            model,
            cutoff_criterion,
        )
        evals = props["cost_hess_evals"]
        history[nfit] = props
        criterion = (
            props["criterion"] if (np.isfinite(evals).all() and (evals > 0).all()) else np.inf
        )
        if verbose and maxscan > 1:
            lowest_criterion = scratch.get("lowest_criterion")
            best = lowest_criterion is None or criterion < lowest_criterion
            if best:
                scratch["lowest_criterion"] = criterion
            print(f"{nfit:7d}  {criterion:10.1f}  {'<---' if best else ''}")
        return criterion

    nfitmax = len(spectrum.freqs) if fcutmax is None else int(spectrum.freqs.searchsorted(fcutmax))
    if nfitmax > nfitmax_hard:
        nfitmax = nfitmax_hard
        warnings.warn(
            "The maximum frequency cutoff is lowered to constrain "
            f"the maximum number of data points in the fit to {nfitmax}.",
            FCutWarning,
            stacklevel=2,
        )
    if nfitmax < nfitmin:
        raise ValueError("Too few data points for fit.")

    if maxscan == 1:
        nfit = nfitmax
        nfits = [nfit]
        compute_criterion(0)
    else:
        nfits = build_xgrid_exp([nfitmin, nfitmax], maxscan)
        rpi_opt(compute_criterion, [0, len(nfits) - 1], mode="min")
        if any(np.isfinite(props["criterion"]) for props in history.values()):
            nfit = min((record["criterion"], key) for key, record in history.items())[1]
        else:
            warnings.warn(
                "Could not find a suitable frequency cutoff. "
                "The resuts for the smallest cutoff are selected.",
                FCutWarning,
                stacklevel=2,
            )
            nfit = nfits[0]

    return Result(spectrum, nfit, dict(sorted(history.items())))


def fit_model_spectrum(
    timestep: float,
    freqs: NDArray[float],
    amplitudes: NDArray[float],
    ndofs: NDArray[int],
    nfit: int,
    model: SpectrumModel,
    cutoff_criterion: CutoffCriterion,
) -> dict[str, NDArray]:
    """Optimize the parameter of a model for a given spectrum.

    The parameters are the attributes of the :py:class:`stacie.cost.LowFreqCost` class,
    except for the ones documented below.

    Parameters
    ----------
    nfit
        The number of low-frequency data points to use in the fit.
    cutoff_criterion
        The criterion function that is minimized to find the best cutoff
        (and thus number of points included in the fit).

    Returns
    -------
    props
        A dictionary with various intermediate results of the cost function calculation,
        computed for the optimized parameters.
        See Notes for details.

    Notes
    -----
    The returned dictionary contains the following items:

    - ``acint``: the estimate of the autocorrelation integral.
    - ``acint_var``: the variance of the estimate of the autocorrelation integral.
    - ``acint_std``: the standard error of the estimate of the autocorrelation integral.
    - ``corrtime``: the estimate of the slowest time scale in the sequences.
    - ``corrtime_var``: the variance of the estimate of the slowest time scale.
    - ``corrtime_std``: the standard error of the estimate of the slowest time scale.
    - ``cost_value``: the cost function value.
    - ``cost_grad``: the cost Gradient vector (if ``deriv>=1``).
    - ``cost_hess``: the cost Hessian matrix (if ``deriv==2``).
    - ``cost_hess_evals``: the Hessian eigenvalues.
    - ``cost_hess_evecs``: the Hessian eigenvectors.
    - ``covar``: the covariance matrix of the parameters.
    - ``criterion``: the value of the criterion whose minimizer determines the frequency cutoff.
    - ``ll``: the log likelihood.
    - ``pars_init``: the initial guess of the parameters.
    - ``pars``: the optimized parameters.
    - ``thetas``: scale parameters for the gamma distribution.
    - ``timestep``: the time step.
    """
    # Maximize likelihood
    model.precondition(timestep, amplitudes[:nfit])
    pars_init = model.guess(freqs[:nfit], amplitudes[:nfit], ndofs[:nfit])
    if not model.valid(pars_init):
        raise AssertionError("Infeasible guess")
    cost = LowFreqCost(timestep, freqs[:nfit], amplitudes[:nfit], ndofs[:nfit], model)
    opt = minimize(
        cost.funcgrad,
        pars_init,
        jac=True,
        hess=cost.hess,
        bounds=model.bounds(),
        method="trust-constr",
        options={"xtol": 1e-10, "gtol": 1e-10},
    )
    props = cost.props(opt.x, 2)

    # Compute the Hessian and its properties.
    evals, evecs = np.linalg.eigh(props["cost_hess"])
    props["cost_hess_evals"] = evals
    props["cost_hess_evecs"] = evecs
    if (evals > 0).all() and np.isfinite(evals).all():
        half = evecs / np.sqrt(evals)
        props["covar"] = np.dot(half, half.T)
    else:
        props["covar"] = np.full_like(props["cost_hess"], np.inf)

    # Derive estimates from model parameters.
    props.update(model.derive_props(props["pars"], props["covar"]))

    # Compute remaining properties and derive the cutoff criterion
    props["pars_init"] = pars_init
    props["freqs_rest"] = freqs[nfit:]
    props["amplitudes_rest"] = amplitudes[nfit:]
    props["kappas_rest"] = 0.5 * ndofs[nfit:]
    props["criterion"] = cutoff_criterion(props)

    # Remove some intermediate properties to reduce the size of the Result object.
    del props["freqs"]
    del props["kappas"]
    del props["amplitudes"]
    del props["freqs_rest"]
    del props["amplitudes_rest"]
    del props["kappas_rest"]

    return props
