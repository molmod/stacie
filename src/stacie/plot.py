# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024-2025 The contributors of the Stacie Python Package.
# See the CONTRIBUTORS.md file in the project root for a full list of contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Plot various aspects of the results of the autocorrelation integral estimate."""

import re

import attrs
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.backends.backend_pdf import PdfPages
from scipy import stats

from .estimate import Result
from .spectrum import Spectrum

__all__ = ("UnitConfig", "plot_results")


@attrs.define
class UnitConfig:
    """Unit configuration for plotting function.

    Note that values are *divided* by their units before plotting.
    """

    acint_symbol: str = attrs.field(default=r"\eta", kw_only=True)
    """The symbol used for the autocorrelation integral."""

    acint_unit_str: str = attrs.field(default="A s", kw_only=True)
    """The text used for the autocorrelation integral unit."""

    acint_unit: float = attrs.field(default=1.0, kw_only=True)
    """The unit of an autocorrelation integral."""

    acint_fmt: str = attrs.field(default=".2e", kw_only=True)
    """The format string for an autocorrelation integral."""

    freq_unit_str: str = attrs.field(default="Hz", kw_only=True)
    """The text used for the frequency unit."""

    freq_unit: float = attrs.field(default=1.0, kw_only=True)
    """The unit of a frequency."""

    time_unit_str: str = attrs.field(default="s", kw_only=True)
    """The text used for a time unit."""

    time_unit: float = attrs.field(default=1.0, kw_only=True)
    """The unit of a frequency."""

    time_fmt: str = attrs.field(default=".2e", kw_only=True)
    """The format string for a time value."""

    sfac: float = attrs.field(default=2.0)
    """The scale factor used for error bars (multiplier for sigma, standard error)."""


def fixformat(s: str) -> str:
    """Replace standard exponential notation with prettier unicode formatting."""

    def repl(match):
        factor = match.group(1)
        exp = str(int(match.group(2)))
        if exp == "0":
            return factor
        return f"${factor}\\times 10^{{{exp}}}$"

    result = re.sub(r"\b([0-9.]+)e([0-9+-]+)\b", repl, s)
    result = re.sub(r"\binf\b", "∞", result)
    return re.sub(r"\bnan\b", "?", result)


def plot_results(
    path_pdf: str, rs: Result | list[Result], uc: UnitConfig | None = None, title: str | None = None
):
    """Generate a multi-page PDF with plots of the autocorrelation integral estimation.

    Parameters
    ----------
    path_pdf
        The PDF file where all the figures are stored.
    rs
        A single ``Result`` instance or a list of them.
        If the (first) result instance has ``spectrum.amplitudes_ref`` set,
        theoretical expectations are included.
        When multiple results instances are given,
        only the first one is plotted in blue.
        All remaining ones are plotted in light grey.
    uc
        The configuration of the units used for plotting.
    """
    # Prepare results list
    if isinstance(rs, Result):
        rs = [rs]

    # Prepare units
    if uc is None:
        uc = UnitConfig()

    with PdfPages(path_pdf) as pdf:
        for r in rs:
            fig, ax = plt.subplots(figsize=(9, 6))
            plot_fitted_spectrum(ax, uc, r)
            if title is not None:
                ax.set_title(title)
            pdf.savefig(fig)
            plt.close(fig)

            if len(r.history) > 1:
                fig, axs = plt.subplots(2, 3, figsize=(9, 6))
                plot_all_models(axs[0, 0], uc, r)
                plot_criterion(axs[1, 0], uc, r)
                plot_uncertainty(axs[0, 1], uc, r)
                plot_residuals(axs[1, 1], uc, r)
                plot_evals(axs[0, 2], uc, r)
                plot_sensitivity(axs[1, 2], uc, r)
                pdf.savefig(fig)
                plt.close(fig)

        if len(rs) > 1:
            if rs[0].spectrum.amplitudes_ref is not None:
                fig, ax = plt.subplots(figsize=(9, 6))
                plot_qq(ax, uc, rs)
                pdf.savefig(fig)
                plt.close(fig)

            fig, ax = plt.subplots(figsize=(9, 6))
            plot_acint_estimates(ax, uc, rs)
            pdf.savefig(fig)
            plt.close(fig)


REF_PROPS = {"ls": "--", "color": "k", "alpha": 0.5}


def plot_spectrum(ax: mpl.axes.Axes, uc: UnitConfig, s: Spectrum, nplot: int | None = None):
    """Plot the empirical spectrum."""
    if nplot is None:
        nplot = s.nfreq
    ax.plot(
        s.freqs[:nplot] / uc.freq_unit,
        s.amplitudes[:nplot] / uc.acint_unit,
        color="C0",
        lw=1,
    )
    _plot_ref_spectrum(ax, uc, s, nplot)
    ax.set_xlim(left=0)
    ax.set_xlabel(f"Frequency [{uc.freq_unit_str}]")
    ax.set_ylabel(f"Spectrum [{uc.acint_unit_str}]")
    ax.set_title("Spectrum")


def _plot_ref_spectrum(ax: mpl.axes.Axes, uc: UnitConfig, s: Spectrum, nplot: int):
    """Plot the reference spectrum."""
    if s.amplitudes_ref is not None:
        ax.plot(
            s.freqs[:nplot] / uc.freq_unit,
            s.amplitudes_ref[:nplot] / uc.acint_unit,
            **REF_PROPS,
        )


FIT_LEFT_TITLE_TEMPLATE = (
    "Spectrum model {model}\n"
    "${uc.acint_symbol}$ = {acint:{uc.acint_fmt}} ± {acint_std:{uc.acint_fmt}}"
    "{acint_unit_str}"
)

FIT_RIGHT_TITLE_TEMPLATE = (
    "$\\tau_\\text{{int}}$ = {corrtime_int:{uc.time_fmt}}"
    " ± {corrtime_int_std:{uc.time_fmt}}"
    "{time_unit_str}"
)

FIT_RIGHT_TITLE_TEMPLATE_EXP = (
    "$\\tau_\\text{{exp}}$ = {corrtime_exp:{uc.time_fmt}}"
    " ± {corrtime_exp_std:{uc.time_fmt}}"
    "{time_unit_str}"
)


def plot_fitted_spectrum(ax: mpl.axes.Axes, uc: UnitConfig, r: Result):
    """Plot the fitted model spectrum."""
    nplot = 2 * r.nfit
    plot_spectrum(ax, uc, r.spectrum, nplot)
    kappas = 0.5 * r.spectrum.ndofs[: r.nfit]
    mean = r.props["thetas"] * kappas
    std = r.props["thetas"] * np.sqrt(kappas)
    freqs = r.spectrum.freqs[: r.nfit]
    ax.plot(freqs / uc.freq_unit, mean / uc.acint_unit, color="C2")
    ax.fill_between(
        freqs / uc.freq_unit,
        (mean - uc.sfac * std) / uc.acint_unit,
        (mean + uc.sfac * std) / uc.acint_unit,
        color="C2",
        alpha=0.3,
        lw=0,
    )
    ax.axvline(r.spectrum.freqs[r.nfit - 1] / uc.freq_unit, ymax=0.1, color="k")
    fields = {
        "uc": uc,
        "model": r.props["model"],
        "acint": r.acint / uc.acint_unit,
        "acint_std": r.acint_std / uc.acint_unit,
        "acint_unit_str": "" if uc.acint_unit_str == "1" else " " + uc.acint_unit_str,
        "corrtime_int": r.corrtime_int / uc.time_unit,
        "corrtime_int_std": r.corrtime_int_std / uc.time_unit,
        "time_unit_str": "" if uc.time_unit_str == "1" else " " + uc.time_unit_str,
    }
    ax.set_title("")
    ax.set_title(fixformat(FIT_LEFT_TITLE_TEMPLATE.format(**fields)), loc="left")
    if "corrtime_exp" in r.props:
        fields["corrtime_exp"] = r.props["corrtime_exp"] / uc.time_unit
        fields["corrtime_exp_std"] = r.props["corrtime_exp_std"] / uc.time_unit
        ax.set_title(
            fixformat(
                FIT_RIGHT_TITLE_TEMPLATE_EXP.format(**fields)
                + "\n"
                + FIT_RIGHT_TITLE_TEMPLATE.format(**fields)
            ),
            loc="right",
        )
    else:
        ax.set_title("\n" + fixformat(FIT_RIGHT_TITLE_TEMPLATE.format(**fields)), loc="right")


def plot_all_models(ax: mpl.axes.Axes, uc: UnitConfig, r: Result):
    """Plot all fitted model spectra (for all tested cutoffs)."""
    for nfit, props in r.history.items():
        kappas = 0.5 * r.spectrum.ndofs[:nfit]
        mean = props["thetas"] * kappas
        freqs = r.spectrum.freqs[:nfit]
        if nfit == r.nfit:
            ax.plot(freqs / uc.freq_unit, mean / uc.acint_unit, color="k", lw=2, zorder=2.5)
        else:
            ax.plot(freqs / uc.freq_unit, mean / uc.acint_unit, color="C2", lw=1, alpha=0.5)
    nplot = min(2 * max(r.history), r.spectrum.nfreq)
    _plot_ref_spectrum(ax, uc, r.spectrum, nplot)
    # Print the number of fitted model spectra in the title to show how many models were tested.
    ax.set_title(f"Model spectra ({len(r.history)} models)")
    ax.set_xlabel(f"Frequency [{uc.freq_unit_str}]")
    ax.set_ylabel(f"Model Spectrum [{uc.acint_unit_str}]")
    ax.set_xscale("log")


def plot_criterion(ax: mpl.axes.Axes, uc: UnitConfig, r: Result):
    """Plot the cutoff criterion as a function of cutoff frequency."""
    freqs = []
    criteria = []
    expected = []
    for nfit, props in sorted(r.history.items()):
        freqs.append(r.spectrum.freqs[nfit - 1])
        criteria.append(props["criterion"])
        expected.append(props.get("criterion_expected", np.nan))
    freqs = np.array(freqs)
    criteria = np.array(criteria)
    expected = np.array(expected)
    mask = np.isfinite(criteria)
    criterion_min = criteria[mask].min()
    criterion_scale = r.props.get("criterion_scale")
    if criterion_scale is None and mask.any():
        criterion_scale = np.median(criteria[mask]) - criterion_min

    if np.isfinite(expected).any():
        ax.plot(freqs / uc.freq_unit, expected, color="C1", lw=1, alpha=0.5, ls="--")
    ax.plot(freqs / uc.freq_unit, criteria, color="C1", lw=1)
    ax.axvline(r.spectrum.freqs[r.nfit - 1] / uc.freq_unit, ymax=0.1, color="k")
    ax.axhline(0, **REF_PROPS)
    ax.set_xlabel(f"Cutoff frequency [{uc.freq_unit_str}]")
    ax.set_ylabel("Criterion")
    ax.set_title("Cutoff criterion")
    ax.set_xscale("log")
    if criterion_scale is not None:
        ax.set_ylim(criterion_min - 0.2 * criterion_scale, criterion_min + 2.5 * criterion_scale)


def plot_uncertainty(ax: mpl.axes.Axes, uc: UnitConfig, r: Result):
    """Plot the autocorrelation integral and uncertainty as a function fo cutoff frequency."""
    freqs = []
    acints = []
    acint_stds = []
    s = r.spectrum
    for nfit, props in sorted(r.history.items()):
        freqs.append(r.spectrum.freqs[nfit - 1])
        acints.append(props["acint"])
        acint_stds.append(props["acint_std"])
    freqs = np.array(freqs)
    acints = np.array(acints)
    acint_stds = np.array(acint_stds)

    ax.plot(freqs / uc.freq_unit, acints / uc.acint_unit, "C3")
    ax.fill_between(
        freqs / uc.freq_unit,
        (acints - uc.sfac * acint_stds) / uc.acint_unit,
        (acints + uc.sfac * acint_stds) / uc.acint_unit,
        color="C3",
        alpha=0.3,
        lw=0,
    )
    s = r.spectrum
    ax.errorbar(
        [r.spectrum.freqs[r.nfit - 1] / uc.freq_unit],
        [r.acint / uc.acint_unit],
        [r.acint_std * uc.sfac / uc.acint_unit],
        marker="o",
        ms=2,
        color="k",
    )
    if s.amplitudes_ref is not None:
        limit = s.amplitudes_ref[0]
        ax.axhline(limit / uc.acint_unit, **REF_PROPS)
    ax.set_xlabel(f"Cutoff frequency [{uc.freq_unit_str}]")
    ax.set_ylabel(f"Autocorrelation integral [{uc.acint_unit_str}]")
    ax.set_xscale("log")


def plot_evals(ax: mpl.axes.Axes, uc: UnitConfig, r: Result):
    """Plot the eigenvalues of the Hessian as a function of the cutoff frequency."""
    freqs = []
    evals = []
    for nfit, props in sorted(r.history.items()):
        freqs.append(r.spectrum.freqs[nfit - 1])
        evals.append(1 / props["cost_hess_evals"])
        if nfit == r.nfit:
            ax.plot([freqs[-1]], [evals[-1]], color="k", marker="o", ms=2, zorder=2.5)
    freqs = np.array(freqs)
    evals = np.array(evals)

    ax.plot(freqs / uc.freq_unit, evals, color="C4")
    ax.set_xlabel(f"Cutoff frequency [{uc.freq_unit_str}]")
    ax.set_ylabel("Covariance eigenvalues")
    ax.set_yscale("log")
    ax.set_xscale("log")


def plot_residuals(ax: mpl.axes.Axes, uc: UnitConfig, r: Result):
    """Plot the normalized residuals between the model and empirical spectra."""
    amplitudes = r.spectrum.amplitudes[: r.nfit]
    kappas = 0.5 * r.spectrum.ndofs[: r.nfit]
    thetas = r.props["thetas"]
    residuals = (amplitudes / thetas - kappas) / np.sqrt(kappas)
    with np.errstate(invalid="ignore"):
        ax.plot(r.spectrum.freqs[: r.nfit] / uc.freq_unit, residuals, color="C0")
        ax.axhline(0, ls="--", lw=1.0, color="k")
    ax.set_title("Normalized residuals")
    ax.set_xlabel(f"Frequency [{uc.freq_unit_str}]")
    ax.set_ylabel("Residual [1]")


def plot_sensitivity(ax: mpl.axes.Axes, uc: UnitConfig, r: Result):
    """Plot the sensitivity of the autocorrelation integral estimate to spectrum amplitudes."""
    ax.plot(r.spectrum.freqs[: r.nfit] / uc.freq_unit, r.props["acint_sensitivity"], color="C5")
    ax.axhline(0, ls="--", lw=1.0, color="k")
    ax.set_title("Sensitivity of the autocorrelation integral to spectrum amplitudes", wrap=True)
    ax.set_xlabel(f"Frequency [{uc.freq_unit_str}]")
    ax.set_ylabel("Sensitivity [1]")


def plot_qq(ax: mpl.axes.Axes, uc: UnitConfig, rs: list[Result]):
    """Make a qq-plot between the predicted and expected distribution of AC integral estimates.

    This plot function assumes the true integral is known.
    """
    cdfs = (np.arange(len(rs)) + 0.5) / len(rs)
    quantiles = stats.norm().ppf(cdfs)
    limit = rs[0].spectrum.amplitudes_ref[0]
    normed_errors = np.array([(r.acint - limit) / r.acint_std for r in rs])
    normed_errors.sort()
    distance = abs(quantiles - normed_errors).mean()
    ax.scatter(quantiles, normed_errors, c="C0", s=3)
    ax.plot([-2, 2], [-2, 2], **REF_PROPS)
    ax.set_xlabel("Normal quantiles [1]")
    ax.set_ylabel("Sorted normalized errors [1]")
    ax.set_title(f"QQ Plot (Wasserstein Distance = {distance:.4f})")


RELERR_TEMPLATE = """\
MRE = {mre:.1f} %
RMSRE = {rmsre:.1f} %
RMSRF = {rmsrf:.1f} %
RMSPRE = {rmspre:.1f} %
"""


def rms(x):
    return np.sqrt((x**2).mean())


def plot_acint_estimates(ax: mpl.axes.Axes, uc: UnitConfig, rs: list[Result]):
    """Plot the sorted autocorrelation integral estimates and their uncertainties."""
    values = np.array([r.acint for r in rs])
    errors = np.array([r.acint_std for r in rs])
    order = values.argsort()
    values = values[order]
    errors = errors[order]
    ax.errorbar(
        np.arange(len(rs)),
        values / uc.acint_unit,
        uc.sfac * errors,
        fmt="o",
        lw=1,
        ms=2,
        ls="none",
    )
    if rs[0].spectrum.amplitudes_ref is not None:
        limit = rs[0].spectrum.amplitudes_ref[0]
        ax.axhline(limit / uc.acint_unit, **REF_PROPS)
        relative_errors = 100 * (values - limit) / values
        mre = relative_errors.mean()
        ax.text(
            0.05,
            0.95,
            RELERR_TEMPLATE.format(
                mre=mre,
                rmsre=rms(relative_errors),
                rmsrf=rms(relative_errors - mre),
                rmspre=rms(errors / values) * 100,
            ),
            transform=ax.transAxes,
            ha="left",
            va="top",
            linespacing=1.5,
        )
    ax.set_xlabel("Rank")
    ax.set_ylabel(f"Mean and uncertainty [{uc.acint_unit_str}]")
    ax.set_title("Autocorrelation Integral")
