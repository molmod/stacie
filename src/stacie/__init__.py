# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024-2025 The contributors of the Stacie Python Package.
# See the CONTRIBUTORS.md file in the project root for a full list of contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""The Stacie package."""

from __future__ import annotations

from .estimate import estimate_acint
from .plot import UnitConfig, plot_results
from .spectrum import compute_spectrum
from .summary import summarize_results

__all__ = (
    "UnitConfig",
    "__version__",
    "__version_tuple__",
    "compute_spectrum",
    "estimate_acint",
    "plot_results",
    "summarize_results",
)


try:
    from ._version import __version__, __version_tuple__
except ImportError:
    __version__ = "0.0.0a-dev"
    __version_tuple__ = (0, 0, 0, "a-dev")
