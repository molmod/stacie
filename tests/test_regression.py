# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Regression tests for typical Stacie workflows."""

import pytest
from path import Path
from stacie.estimate import estimate_acfint
from stacie.plot import plot
from stacie.spectrum import Spectrum
from stacie.zarr import load

DOUBLE_NAMES = ["double1", "double2"]
WHITE_NAMES = ["white1", "white2"]
NAME_LISTS = DOUBLE_NAMES, WHITE_NAMES
ALL_NAMES = [j for i in NAME_LISTS for j in i]


def plot_test_result(prefix, res):
    dn_out = Path("tests/outputs")
    dn_out.makedirs_p()
    path_pdf = dn_out / f"{prefix}.pdf"
    plot(path_pdf, res)


def register_result(regtest, res):
    with regtest:
        print(f"ACF Int  = {res.props['acfint']:.4f} ± {res.props['acfint_std']:.4f}")
        print(f"Tau tail = {res.props['corrtime_tail']:.4f} ± {res.props['corrtime_tail_std']:.4f}")
        print(f"Log lh   = {res.props['ll']:.5e}")
        print("---")


def check_noscan_single(regtest, spectrum, prefix, fcut=0.005):
    res = estimate_acfint(spectrum, fcut=fcut, maxscan=1)
    register_result(regtest, res)
    plot_test_result(prefix, res)


@pytest.mark.parametrize("name", ALL_NAMES)
def test_noscan(regtest, name):
    spectrum = load(f"tests/inputs/spectrum_{name}.zip", Spectrum)
    check_noscan_single(regtest, spectrum, f"noscan_{name}")


@pytest.mark.parametrize(("name", "fcut"), [("white2", 0.008), ("double1", 0.05)])
def test_noscan_fail(regtest, name, fcut):
    spectrum = load(f"tests/inputs/spectrum_{name}.zip", Spectrum)
    check_noscan_single(regtest, spectrum, f"noscan_{name}_fail", fcut=fcut)


@pytest.mark.parametrize("names", NAME_LISTS)
def test_noscan_multi(regtest, names):
    res = []
    for name in names:
        spectrum = load(f"tests/inputs/spectrum_{name}.zip", Spectrum)
        r = estimate_acfint(spectrum, fcut=0.005, maxscan=1)
        register_result(regtest, r)
        res.append(r)
    plot_test_result("noscan_multi", res)


def check_scan_single(regtest, spectrum, prefix):
    res = estimate_acfint(spectrum, fcut=0.01, maxscan=10)
    register_result(regtest, res)
    plot_test_result(prefix, res)


@pytest.mark.parametrize("name", ALL_NAMES)
def test_scan(regtest, name):
    spectrum = load(f"tests/inputs/spectrum_{name}.zip", Spectrum)
    check_scan_single(regtest, spectrum, f"scan_{name}")


@pytest.mark.parametrize("names", NAME_LISTS)
def test_scan_multi(regtest, names):
    res = []
    for name in names:
        spectrum = load(f"tests/inputs/spectrum_{name}.zip", Spectrum)
        r = estimate_acfint(spectrum, fcut=0.01, maxscan=10)
        register_result(regtest, r)
        res.append(r)
    plot_test_result("scan_multi", res)
