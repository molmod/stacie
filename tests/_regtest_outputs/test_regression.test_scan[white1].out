ACF Int  = 140.5952 ± 6.4655
Tau tail = 275.8127 ± 169.9681
Log lh   = -6.10718e+01
---
